mission_control/navigator_missions/nav_missions/start_gate.py

```diff
@@ -1,66 +1,277 @@
+    #rand_2 = Buoy([33, -8, 0], "green")
+
+    buoys = [right, left] #[rand_1, right, rand_2, left]
+
+    return buoys
+
+
+def get_path(buoy_l, buoy_r):
+    # Vector between the two buoys
+    between_vector = buoy_r.position - buoy_l.position
+
+    # Rotate that vector to point through the buoys
+    r = trns.euler_matrix(0, 0, np.radians(90))[:3, :3]
+    direction_vector = r.dot(between_vector)
+    direction_vector /= np.linalg.norm(direction_vector)
+
+    return between_vector, direction_vector
+
+
+return_with = defer.returnValue
 
 @txros.util.cancellableInlineCallbacks
 def main(navigator):
     '''
-    Once we see the gate - move to a point 10m in front of the gate.
-    When we get there, make another observation of the gate to make
-        sure we are lined up and then go through.
-
+    This assumes we have ALL FOUR buoys
     '''
-    navigator.change_wrench("autonomous")
-    # Only try to get the point 5 times
-    for _try in range(5):
-        print "Waiting for points..."
-        m = yield navigator.vision_request("start_gate")
-        print m
+    result = navigator.fetch_result()
+
+    #navigator.change_wrench("autonomous")
+    buoys = get_buoys()
+    #serv = navigator.nh.get_service_client("/database/single", ObjectDBSingleQuery)
+    #req = ObjectDBSingleQueryRequest()
+    #req.name = 'l_gate'
+    #left = yield serv(req)
+    #req.name = 'r_gate'
+    #right = yield serv(req)
+    #buoys = [Buoy.from_srv(left), Buoy.from_srv(right)]
+    print buoys
+    # Display points of the buoys]
+    points = [navigator_tools.numpy_to_point(b.position) for b in buoys]
+
+    pose = yield navigator.tx_pose
+    buoys = sorted(buoys, key=lambda buoy: buoy.distance(pose[0]))
+
+    # print pose
+    # print buoys
+
+    print "\nSTART_GATE: Be advised... Doing checks"
+
+    # See if the closest two buoys are the right colors ============================================
+    if not (buoys[0].color in ['green', 'red'] and buoys[1].color in ['green', 'red']):
+        print "START_GATE: Error 1 - Not the right colors. ({},{})".format(buoy[0].color,
+                                                                           buoy[1].color)
+        result.success = False
+        #return_with(result)
+
+    # Okay, now we need to see if the buoys are approx 10m apart ===================================
+    nominal_dist = 10  # m
+    variance = 5  #m
+    dist_btwn_buoys = np.linalg.norm(buoys[0].position[:2] - buoys[1].position[:2])
+    if not (nominal_dist - variance < dist_btwn_buoys < nominal_dist + variance):
+        print "START_GATE: Error 2 - Not the right distance apart. {}".format(dist_btwn_buoys)
+        #result.success = False
+        #return_with(result)
+
+    # Made it this far, make sure the red one is on the left and green on the right ================
+    #t = yield navigator.tf_listener.get_transform("/base_link", "/enu", navigator.nh.get_time())
+    class t(object):
```

Comment:
Simulation - not having to have a transform publisher running to get things working.
But you're right, I will revise to remove all the simulation stuff so it's runnable on the boat.